pkg/rosa/oidcconfigs/oidcconfigs.go

```diff
@@ -0,0 +1,217 @@
+	bucketUrl := fmt.Sprintf("https://%s.s3.%s.amazonaws.com", bucketName, region)
+	privateKey, publicKey, err := CreateKeyPair()
+	if err != nil {
+		return OidcConfigInput{}, fmt.Errorf("There was a problem generating key pair: %s", err)
+	}
+	privateKeyFilename := fmt.Sprintf("%s.key", privateKeySecretName)
+	discoveryDocument := GenerateDiscoveryDocument(bucketUrl)
+	jwks, err := BuildJSONWebKeySet(publicKey)
+	if err != nil {
+		return OidcConfigInput{}, fmt.Errorf("There was a problem generating JSON Web Key Set: %s", err)
+	}
+	return OidcConfigInput{
+		BucketName:           bucketName,
+		IssuerUrl:            bucketUrl,
+		PrivateKey:           privateKey,
+		PrivateKeyFilename:   privateKeyFilename,
+		DiscoveryDocument:    discoveryDocument,
+		Jwks:                 jwks,
+		PrivateKeySecretName: privateKeySecretName,
+	}, nil
+}
+
+func GenerateBucketName(userPrefix string) (string, error) {
+	randomLabel := utils.RandomLabel(defaultLengthRandomLabel)
+	bucketName := fmt.Sprintf("%s-%s", defaultPrefixForConfiguration, randomLabel)
+	if userPrefix != "" {
+		bucketName = fmt.Sprintf("%s-%s", userPrefix, bucketName)
+	}
+	if !IsValidBucketName(bucketName) {
+		return "", fmt.Errorf("The bucket name '%s' is not valid", bucketName)
+	}
+
+	return bucketName, nil
+}
+
+const (
+	bucketNameRegex = "^[a-z][a-z0-9\\-]+[a-z0-9]$"
+)
+
+func IsValidBucketName(bucketName string) bool {
+	if bucketName[0] == '.' || bucketName[len(bucketName)-1] == '.' {
+		return false
+	}
+	if strings.HasPrefix(bucketName, "xn--") {
+		return false
+	}
+	if strings.HasSuffix(bucketName, "-s3alias") {
+		return false
+	}
+	if match, _ := regexp.MatchString("\\.\\.", bucketName); match {
+		return false
+	}
+	// We don't support buckets with '.' in them
+	match, _ := regexp.MatchString(bucketNameRegex, bucketName)
+	return match
+}
+
+func CreateKeyPair() ([]byte, []byte, error) {
+	bitSize := 4096
+
+	// Generate RSA keypair
+	privateKey, err := rsa.GenerateKey(rand.Reader, bitSize)
+	if err != nil {
+		return nil, nil, fmt.Errorf("Failed to generate private key: %v", err)
+	}
+	encodedPrivateKey := pem.EncodeToMemory(&pem.Block{
+		Type:    "RSA PRIVATE KEY",
+		Headers: nil,
+		Bytes:   x509.MarshalPKCS1PrivateKey(privateKey),
+	})
+
+	// Generate public key from private keypair
+	pubKeyBytes, err := x509.MarshalPKIXPublicKey(&privateKey.PublicKey)
+	if err != nil {
+		return nil, nil, fmt.Errorf("Failed to generate public key from private: %v", err)
+	}
+	encodedPublicKey := pem.EncodeToMemory(&pem.Block{
+		Type:    "PUBLIC KEY",
+		Headers: nil,
+		Bytes:   pubKeyBytes,
+	})
+
+	return encodedPrivateKey, encodedPublicKey, nil
+}
+
+const (
+	discoveryDocumentTemplate = `{
+	"issuer": "%s",
+	"jwks_uri": "%s/keys.json",
+	"response_types_supported": [
+		"id_token"
+	],
+	"subject_types_supported": [
+		"public"
+	],
+	"id_token_signing_alg_values_supported": [
+		"RS256"
+	],
+	"claims_supported": [
+		"aud",
+		"exp",
+		"sub",
+		"iat",
+		"iss",
+		"sub"
```

Comment:
should sub be here twice?

Reply:
huh, I actually only copied. So I guess it's always been a dupe